src/App.tsx

```diff
@@ -1,24 +1,214 @@
+  const [hideNotification, setHideNotification] = useState(true);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isFocused, setIsFocused] = useState(true);
+  const [allTodosCompleted, setAllTodosCompleted] = useState(false);
+  const [makeTodosComplete, setMakeTodosComplete]
+  = useState(MakeTodosCompleted.begin);
+  const [clearCompleted, setClearCompleted] = useState(false);
+
+  useEffect(() => {
+    const filteredTodos = todos.filter((todo) => todo.completed);
+
+    setQuantity(todos.length - filteredTodos.length);
+  }, [todos]);
+
+  useEffect(() => {
+    setErrorMessage('');
+
+    todosService
+      .getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      });
+  }, []);
+
+  const filteredTodos = useMemo(() => {
+    const filteredByTodos = todos.length
+      ? todos.filter((todo) => {
+        switch (filterBy) {
+          case Filter.active:
+            return !todo.completed;
+          case Filter.completed:
+            return todo.completed;
+          default:
+            return true;
+        }
+      })
+      : [];
+
+    return filteredByTodos;
+  }, [filterBy, todos]);
+
+  const deleteTodo = (todoId: number) => {
+    setErrorMessage('');
+    setIsFocused(false);
+
+    return todosService
+      .deleteTodo(todoId)
+      .then(() => {
+        setTodos((currentTodos) => currentTodos
+          .filter((todo) => todo.id !== todoId));
+      })
+      .catch((error) => {
+        setErrorMessage('Unable to delete a todo');
+        throw error;
+      })
+      .finally(() => {
+        setIsFocused(true);
+      });
+  };
+
+  const createNewTodo = ({ title, completed, userId }: Omit<Todo, 'id'>) => {
+    setErrorMessage('');
+
+    return todosService
+      .addTodo({ title, completed, userId })
+      .then((newTodo: Todo) => {
+        setTodos((currentTodos) => [...currentTodos, newTodo]);
+      })
+      .catch((error) => {
+        setErrorMessage('Unable to add a todo');
+        throw error;
+      });
+  };
+
+  const updateTodo = (updatedTodo: Todo) => {
+    setErrorMessage('');
+
+    return todosService
+      .updateTodo(updatedTodo)
+      .then((todo) => {
+        setTodos((currentTodos) => {
+          const newTodos = [...currentTodos];
+          const index = newTodos.findIndex(
+            (post) => post.id === updatedTodo.id,
+          );
+
+          newTodos.splice(index, 1, todo);
+
+          return newTodos;
+        });
+      })
+      .catch((error) => {
+        setErrorMessage('Unable to update a todo');
+        throw error;
+      });
+  };
+
+  const searchCompletedTodos = () => {
+    const hasCompletedTodos = filteredTodos.some((todo) => todo.completed);
+
+    setHasCompleted(hasCompletedTodos);
+  };
+
+  useEffect(() => {
+    const hasCompletedTodos = filteredTodos.some((todo) => todo.completed);
+
+    setHasCompleted(hasCompletedTodos);
+  }, [filteredTodos, hasCompleted]);
```

Comment:
`hasCompleted` it is the same as `quantity === 0`, isn't it?

Reply:
if so, then you don't need this effect and this state at all, but I'm not sure, please check

Reply:
quantity shows us how many todos aren't completed. We use it to show this amount in the left bottom of app.
state hasCompleted gives us possibility to disable button clear all completed